examples/js/postprocessing/SSAOPass.js

```diff
@@ -1,190 +1,375 @@
-	this.scene2 = scene;
+	this.depthRenderMaterial = new THREE.ShaderMaterial( {
+		defines: Object.assign( {}, THREE.SSAODepthShader.defines ),
+		uniforms: THREE.UniformsUtils.clone( THREE.SSAODepthShader.uniforms ),
+		vertexShader: THREE.SSAODepthShader.vertexShader,
+		fragmentShader: THREE.SSAODepthShader.fragmentShader,
+		blending: THREE.NoBlending
+	} );
+	this.depthRenderMaterial.uniforms[ 'tDepth' ].value = this.beautyRenderTarget.depthTexture;
+	this.depthRenderMaterial.uniforms[ 'cameraNear' ].value = this.camera.near;
+	this.depthRenderMaterial.uniforms[ 'cameraFar' ].value = this.camera.far;
+
+	// material for rendering the content of a render target
+
+	this.copyMaterial = new THREE.ShaderMaterial( {
+		uniforms: THREE.UniformsUtils.clone( THREE.CopyShader.uniforms ),
+		vertexShader: THREE.CopyShader.vertexShader,
+		fragmentShader: THREE.CopyShader.fragmentShader,
+		transparent: true,
+		depthTest: false,
+		depthWrite: false,
+		blendSrc: THREE.DstColorFactor,
+		blendDst: THREE.ZeroFactor,
+		blendEquation: THREE.AddEquation,
+		blendSrcAlpha: THREE.DstAlphaFactor,
+		blendDstAlpha: THREE.ZeroFactor,
+		blendEquationAlpha: THREE.AddEquation
+	} );
 
-	//Depth material
-	this.depthMaterial = new THREE.MeshDepthMaterial();
-	this.depthMaterial.depthPacking = THREE.RGBADepthPacking;
-	this.depthMaterial.blending = THREE.NoBlending;
+	//
 
-	//Depth render target
-	this.depthRenderTarget = new THREE.WebGLRenderTarget( this.width, this.height, { minFilter: THREE.LinearFilter, magFilter: THREE.LinearFilter } );
-	//this.depthRenderTarget.texture.name = 'SSAOShader.rt';
+	this.quadCamera = new THREE.OrthographicCamera( - 1, 1, 1, - 1, 0, 1 );
+	this.quadScene = new THREE.Scene();
+	this.quad = new THREE.Mesh( new THREE.PlaneBufferGeometry( 2, 2 ), null );
+	this.quadScene.add( this.quad );
 
-	//Shader uniforms
-	this.uniforms[ 'tDepth' ].value = this.depthRenderTarget.texture;
-	this.uniforms[ 'size' ].value.set( this.width, this.height );
-	this.uniforms[ 'cameraNear' ].value = this.camera2.near;
-	this.uniforms[ 'cameraFar' ].value = this.camera2.far;
+};
 
-	this.uniforms[ 'radius' ].value = 4;
-	this.uniforms[ 'onlyAO' ].value = false;
-	this.uniforms[ 'aoClamp' ].value = 0.25;
-	this.uniforms[ 'lumInfluence' ].value = 0.7;
+THREE.SSAOPass.prototype = Object.assign( Object.create( THREE.Pass.prototype ), {
 
-	//Setters and getters for uniforms
+	constructor: THREE.SSAOPass,
 
-	Object.defineProperties( this, {
+	render: function ( renderer, writeBuffer /*, readBuffer, delta, maskActive */ ) {
 
-		radius: {
-			get: function () {
+		// render beauty and depth
 
-				return this.uniforms[ 'radius' ].value;
+		renderer.setClearColor( 0x000000 );
+		renderer.render( this.scene, this.camera, this.beautyRenderTarget, true );
 
-			},
-			set: function ( value ) {
+		// render normals
 
-				this.uniforms[ 'radius' ].value = value;
+		this.renderOverride( renderer, this.normalMaterial, this.normalRenderTarget, 0x7777ff, 1.0 );
```

Comment:
This is changing the background color and not being reset to the original value, apparently.

Reply:
`.renderOverride()` should actually reset the original state at the end of the method.

Reply:
To demonstrate, in `webgl_postprocessing_ssao.html`, remove the scene background:
```javascript
//scene.background = new THREE.Color( 0xa0a0a0 );
```
and add a call to render
```javascript
effectComposer.render();
renderer.render( scene, camera );
```

Reply:
I think I've found the problem right [here](https://github.com/mrdoob/three.js/blob/e94fd574f52bd8d2e6f03e47913d577806c0f5ad/examples/js/postprocessing/SSAOPass.js#L262). `WebGLRenderer.getClearColor()` returns no copy of the color. If the color is [later](https://github.com/mrdoob/three.js/blob/e94fd574f52bd8d2e6f03e47913d577806c0f5ad/examples/js/postprocessing/SSAOPass.js#L275) modified, it also affects the original value. A simple fix would be:
```js
this.originalClearColor.copy( renderer.getClearColor() );
```
and later
```js
renderer.setClearColor( this.originalClearColor );
```

Reply:
Would #15212 have been helpful in this case, or are there problems with that, too?

Reply:
Not sure. This seems app-level logic to me. Maybe it's better if the app saves and restores these values. Especially if you need to save/restore more than just a single pair of clear color and alpha values.

Reply:
>This seems app-level logic to me.

Granted. I just see the same patterns used over and over...

Reply:
If these methods are added, users could ask for the possibility to save multiple states (think of vertex array objects that represent binding states) and restore/load them with a ID or a name. I think stuff like that would bloat the API of `WebGLRenderer`.

Reply:
>If these methods are added, users could ask for the possibility...

Anything could happen. Sorry, I'm not buying that as a very compelling argument.

>This seems app-level logic to me.

That's a reasonable argument.